[RFC][Validator] Allow to set the translation domain or disable translations on a per-constraint level

mpdude opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | 
| License       | MIT

This PR builds upon the work in #48852 and makes it possible to specify the translation domain for validation constraints on a per-constraint level. This can happen either when using the constraint as an attribute, or when creating constraint instances e. g. when building a form.

```php
   $formBuilder
            ->add('some_field', TextType::class, [
                'label' => 'Something',
                'constraints' => [new NotBlank(['message' => 'standard.notblank.message', 'translation_domain' => 'my_special_domain'])],
            ])
            ->add('other_field', TextType::class, [
                'label' => 'Elsething',
                'constraints' => [new NotBlank(['message' => 'Use this text as-is, do not attempt to translate', 'translation_domain' => false])],
            ]);
```

The setting works similar to the `translation_domain` option for form types. Setting it to `false` disables translations (for that constraint) altogether.

#### TODO

- [ ] Add documentation PR
- [ ] Apply enhancement to all constraints once we agree on the basics; for now, it's just `NotBlank`
- [ ] Add tests
- [ ] Update src/Symfony/Component/Translation/Extractor/Visitor/ConstraintVisitor.php

mpdude committed:
Allow to set the translation domain or disable translations on a per-constraint level

mpdude committed:
Add a CHANGELOG notice

mpdude committed:
Update two tests

nicolas-grekas commented on src/Symfony/Component/Validator/Violation/ConstraintViolationBuilder.php:
```diff
@@ -83,6 +83,16 @@ public function disableTranslation(): static
+    /**
+     * @return $this
+     */
+    public function setOrDisableTranslationDomain(string|false|null $translationDomain): static
```
```suggestion
    public function setTranslationDomain(string|false|null $translationDomain): static
```

mpdude replied:
There is already such a method:

`public function setTranslationDomain(string $translationDomain): static`

We can't pass it `null|false`, but I did not want to put extra checks into each validator to call e. g. `\Symfony\Component\Validator\Violation\ConstraintViolationBuilder::disableTranslation()` in that case. That's why I added a new, all-purpose method.

nicolas-grekas reviewed (commented):
Is there really a use case for that?

mpdude commented:
We have monitoring in place that reports certain errors from production systems, and the translator warnings are among those.

Sometimes, you don't want to go through the hassle of maintaining translation catalogues and put the plain text into the code directly (single-lang apps).

In that case, it would be nice if translations could be disabled.

nicolas-grekas commented:
How do you deal with that at the moment?
I feel like the proposed strategy has too much impact for achieving your use case.
Any other idea that'd be less invasive?

ro0NL commented on src/Symfony/Component/Validator/Constraints/NotBlank.php:
```diff
@@ -31,6 +31,9 @@ class NotBlank extends Constraint
     ];
 
     public string $message = 'This value should not be blank.';
```
i'd build upon `string|TranslatableMessage $message`

mpdude commented:
> Any other idea that'd be less invasive?

... or do you mean "achieves the same without a repeating code pattern in all constraints and validators"?

Full ack, I do not fancy putting the same change into lots of classes over and over again.

nicolas-grekas commented:
Thanks for the hints. I've no better idea so I feel like this might be good to me.
I'm just wondering: would it make sense to allow defining the translation domain by contraint's FQCN in config/packages/validations.yaml? This way one could say that all `#[SomeConstraint]` messages should to some domain?

nicolas-grekas commented:
>... or do you mean "achieves the same without a repeating code pattern in all constraints and validators"?

That's what I meant yes.
Maybe my config idea could be enough?

```yaml
validation:
  translation_domains:
    SomeConstraint: some-domain
```

mpdude commented:
Not sure – assume you have two bundles that build a `FormType` each. And in those `FormTypes`, they both use the `NotBlank` constraint with message-IDs.

To me, it would make sense if both bundles could keep their respective validation/constraint message in their "own" translation domain, as suggested by the [best practice](https://symfony.com/doc/current/bundles/best_practices.html#translation-files). 

That would not be possible with a config setting as suggested, or am I missing something?

Would you be fine with a solution that involves something with traits to plug-in the `translation_domain` into all the constraints?

Still, that does not allow to pass it as a constructor argument. 🤔

Also, since the validator class calls `$this->context->buildViolation()` to create the `ConstraintViolationBuilder`, but never passes it somewhere else, and ends with calling `addViolation()` without ever passing the `Constraint` instance, I am afraid there is nothing we can do with the current code design to centralize`translation_domain` handling.

Maybe GitHub Copilot can pull it off for us 😹?

mpdude commented:
@ro0NL You made a comment about `TranslatableMessage` support; did you withdraw that for good reasons? Probably makes the change even bigger.

ro0NL commented:
If we want to customize per constraint and per instance we could widen the message option to `string|TranslatableInterface`, which provides all flexibility. But i believe it's a BC break.

welcoMattic commented:
Adding this feature will break the PhpAstExtractor as it is. Actually, to extract translation keys from Constraints, it leverages https://github.com/symfony/symfony/blob/7.3/src/Symfony/Component/Translation/Extractor/Visitor/ConstraintVisitor.php, which is looking keys and extract them in `validators` domain by default.

Introducing `translation_domain` option here means to update the ConstraintVisitor to handle the `translation_domain` option value to extract keys in the correct domain (or `validators` by default as it is now).

IIUC, the main reason of this feature is:

> https://github.com/symfony/symfony/pull/48852 put the logic in place to change or disable the translation domain for individual (project-specific) validators, but does not make this feature available for the built-in constraints.

But I'm wondering why do you need to change the translation domain of built-in constraints? Did I miss something?

mpdude commented:
@welcoMattic thank you for the heads up! If I get you right, I'll need to update the `src/Symfony/Component/Translation/Extractor/Visitor/ConstraintVisitor.php` as well.

It's not about changing the translation domain for built-in messages; it's great that Symfony ships with translations for those!

My use case is about using the constraints like `NotBlank` in custom form types, potentially shipped in bundles, with custom validation error messages. For those, I'd like to be able to control the message catalogue. See my initial comment for the example.

Does that make sense?

Still, I have no better idea yet how to approach this apart from making the same kind of change in all constraints. Anyone else?

mpdude committed:
Add a test to check that the translation domain is passed on

nicolas-grekas commented:
So, instead of spreading this everywhere, maybe we could just update the NotBlank one?
And maybe a few more constraints where chancing the default message makes sense?

mpdude commented:
I've fixed tests and added an example of what a test case might look like.

The complete change ain't gonna be fun – so maybe @nicolas-grekas can give me a :+1: that you would indeed merge it if I go the whole way of adding it? 😉

mpdude commented:
So you'd prefer changing only a few constraints over consistency of usage?

nicolas-grekas commented:
I do yes. Consistency is a superficial thing compared to code complexity.

nicolas-grekas commented:
Up to finish this PR @mpdude ?

mpdude commented:
I checked out the codebase where I'd benefit from this change and was daunted due to the number of validators actually used 🙈.

Let me think through again whether I really want to go on this journey.
